### ISSUE TITLE
Cannot inject a custom principal into unauthenticated request

### ISSUE THREAD

**Issue by Gaudon** [dropwizard/dropwizard#2041]
**The Goal**
Use the @Auth annotation in resource methods to validate that a specific **authenticated** user has explicit permission to **authorize** a particular resource. 

This involves accepting a token from the client, mapping that token to the user to which it belongs and ensuring that the identified user has the required permission to view the requested endpoint. 

This seems pretty straight forward in my mind, but I'm having some issues integrating this in the latest version of DropWizard as a first time user, with most online resources no longer being relevant to Jersey() 2.0.

**Console:**

> ERROR
 [2017-05-02 11:34:41,302] io.dropwizard.jersey.errors.LoggingExceptionMapper: Error handling a request: 72ab3581d1fd9815. 
java.lang.IllegalStateException: Cannot inject a custom principal into unauthenticated request

**Application Setup**

```java
DashboardAuthenticator authenticator = new UnitOfWorkAwareProxyFactory(hibernateBundle).create(DashboardAuthenticator.class, new Class[]{SessionDAO.class, UserDAO.class}, new Object[]{sessionDAO, userDAO});

DashboardAuthorizer authorizer = new DashboardAuthorizer(sessionDAO);

DashboardAuthFilter authFilter = new DashboardAuthFilter(authenticator, authorizer);

environment.jersey().register(new AuthDynamicFeature(authFilter));
environment.jersey().register(new AuthValueFactoryProvider.Binder<>(DashboardCredentials.class));
```

**Flow Of Request**

Path: http://localhost:8080/students

DashboardCredentials.java (Implements Principal) - Constructor
DashboardAuthenticator(SessionDAO, UserDAO).authenticate() Method (Does authentication of user based on provided Authentication header token)
Exception Thrown Here.

**StudentsResource [Sample]**

```java
@Path("/students")
@PermitAll
@Produces(MediaType.APPLICATION_JSON)
public class StudentsResource {

    private final StudentDAO studentDAO;

    public StudentsResource(StudentDAO studentDAO) {
        this.studentDAO = studentDAO;
    }
    @GET
    @UnitOfWork
    public List<Student> getAllStudents(@Auth DashboardCredentials credentials) {
        // Validate Authorization Here
        return studentDAO.getAll();
    }
```

**Notes**
I have tried several iterations and modifications of the above. I cannot seem to piece together what DropWizard expects and how this works. Every example online only uses BasicAuthentication in the most simple way and isn't painting a clear picture as to how to accomplish this.

**Comment by nickbabcock** [comment]
Dropwizard also ships with a [`OAuthCredentialAuthFilter`](https://github.com/dropwizard/dropwizard/blob/223be464b1e58ac8e85137c71bbb541bed760403/dropwizard-auth/src/main/java/io/dropwizard/auth/oauth/OAuthCredentialAuthFilter.java), which you can take inspiration from.

**Comment by Gaudon** [comment]
I've taken a look at most of, if not all the available resources on DropWizard-example as well as the source here on github. Including the OAuthCredentialAuthFilter mentioned.

It could be that I'm an idiot and doing something glaringly incorrectly, but if so, I just can't seem to find it or understand what it is.

**Comment by nickbabcock** [comment]
Yeah, nothing erroneous stands out to me.

Some of the things I'd try for:

- Does `DashboardCredentials` implement Principal?
- Remove Hiberate
- Remove authorization
- Return a dummy authentication object
- Ensure the filter is firing and works as expected

**Comment by Gaudon** [comment]
Most of these are things I also thought to check and can give you direct answers for. 

- Does DashboardCredentials implement Principal? **Yes.**
- Remove Hibernate. **Haven't tried this, but I don't want to use the framework without ORM support**
- Remove authorization. **This effectively disables the issue. Yes the API works as expected, but without any sort of authentication or authorization**
- Return a dummy authentication object. **I'm not sure what good this would do. I'm unable to return anything from authorize() and have a non-null value in the @Auth annotations.**
- Ensure the filter is firing and works as expected. **The filter does fire, validate the data with respect to the database as expected.**

**Comment by nickbabcock** [comment]
The suggestions are temporary to help narrow the problem to determine whether it's a user or a dropwizard issue.

**Comment by Gaudon** [comment]
I will remove hibernate and see how much breaks. If it's possible to test it easily without Hibernate I will update my comment and post back.

**Comment by Gaudon** [comment]
In an attempt to remove Hibernate and make authentication not @unitofwork aware I did the following.
```
// Authentication  
//DashboardAuthenticator authenticator = new UnitOfWorkAwareProxyFactory(hibernateBundle).create(DashboardAuthenticator.class, new Class[]{SessionDAO.class, UserDAO.class}, new Object[]{sessionDAO, userDAO});
DashboardAuthenticator authenticator = new DashboardAuthenticator(sessionDAO, userDAO);  
DashboardAuthorizer authorizer = new DashboardAuthorizer(sessionDAO);
DashboardAuthFilter authFilter = new DashboardAuthFilter(authenticator, authorizer);

environment.jersey().register(new AuthDynamicFeature(authFilter));
environment.jersey().register(new AuthValueFactoryProvider.Binder<>(DashboardCredentials.class));
```

**DashboardAuthenticator.java**
```
@Override
//@UnitOfWork
public Optional<DashboardCredentials> authenticate(Object credentials) throws AuthenticationException {
        DashboardCredentials dbCredentials = (DashboardCredentials)credentials;
        dbCredentials.setToken("ABCDEF");
        dbCredentials.setUserId(45);

        return Optional.of(dbCredentials);
}
```

Yields the following error as before:

> ERROR [2017-05-02 15:40:33,637] io.dropwizard.jersey.errors.LoggingExceptionMapper: Error handling a request: 3535da080f38cdef
! java.lang.IllegalStateException: Cannot inject a custom principal into unauthenticated request

**Comment by nickbabcock** [comment]
So, I'm still thinking, but is there any reason why DashboardAuthenticator doesn't implement `Authenticator<DashboardCredentials, DashboardUser>`

**Comment by Gaudon** [comment]
Not entirely, no. It certainly could. 

The DashboardCredentials for all intents and purposes only needs to contain the Token to validate it's an active token and inversely find the associated user to which it belongs. 

This in turn could be used to populate the suggested DashboardUser class. 

It's a better design choice, and I'm changing it to do that just now. 

I don't think it's related or the root cause however to the issue I'm experiencing.

**Comment by Gaudon** [comment]
```
public class DashboardAuthenticator implements Authenticator<DashboardCredentials, DashboardUser> {
    private SessionDAO sessionDAO;
    private UserDAO userDAO;

    public DashboardAuthenticator(SessionDAO sessionDAO, UserDAO userDAO) {
        this.sessionDAO = sessionDAO;
        this.userDAO = userDAO;
    }

   @Override
   @UnitOfWork
   public Optional<DashboardUser> authenticate(DashboardCredentials credentials) throws AuthenticationException {
        Session session = sessionDAO.findSession(credentials.getToken());
        ...
        User user = userDAO.findByUserId(session.getUserId());
        DashboardUser dashboardUser = new DashboardUser(user.getEmail(), user.getId());
        return Optional.of(dashboardUser);
   }
```

**Comment by nickbabcock** [comment]
You see [AuthFilter#authenticate](https://github.com/dropwizard/dropwizard/blob/9a311914295a5cc81a59f6dc5981286a57b0284f/dropwizard-auth/src/main/java/io/dropwizard/auth/AuthFilter.java#L130-L169) how it calls `setSecurityContext`? Your custom filter needs to either call the super method or set security context some other way.

**Comment by Gaudon** [comment]
You are absolutely correct, and that completely fixed it.

I don't know what "Authentication Scheme" as a string should be in this context, but I will look into what it's used for. 

In the mean time I'll update the above class in case other people have this issue.

Very much appreciate the help. I tried to find a donation link off your github page links to buy you a beer. I failed to find any such link! Thanks all the same.
